keps/sig-node/4781-kublet-restart-pod-status/README.md

```diff
@@ -0,0 +1,378 @@
+
+It's essential to implement strategies to ensure that the service states accurately reflect their operational status, even during kubelet interruptions.
+
+### Goals
+
+- Ensure consistency in container start and ready states across kubelet restarts.
+- Minimize unnecessary service disruptions caused by temporary ready state changes.
+
+### Non-Goals
+
+- If the kubelet fails to renew its lease beyond the nodeMonitorGracePeriod due to an excessively long restart interval, the Ready status of the containers in the pods on the node will be set to false. In this situation, we should not manually set the Ready status back to true. Instead, it should remain false, waiting for the probe to execute again and restore it.
+- Modify the fundamental logic of how readiness probes work.
+
+## Proposal
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+As a user of Kubernetes, I want the container's Ready state to remain consistent across kubelet restarts so that my services do not experience unnecessary downtime.
+However, currently, a kubelet restart causes a brief "Not Ready" storm, where the state of all Pods is set to Not Ready, impacting the availability of my services.
+
+### Risks and Mitigations
+
+##### Inconsistency with other Kubernetes components
+If other parts of Kubernetes (e.g., the API server, controllers) expect certain behavior regarding container readiness states, these changes might cause inconsistencies.
+
+##### Delayed Health Check Updates
+By preserving the old state without immediate health checks, there is a delay in recognizing containers that have become unhealthy during or after kubelet's downtime. Services relying on Pod readiness for service discovery might continue directing traffic to Pods with containers that are no longer healthy but are still reported as Ready.
+
+## Design Details
+
+We will be adding a deprecated feature gate: `ChangeContainerStatusOnKubeletRestart`. This feature gate will be disabled by default. When disabled, the Kubelet will not change container statuses after a restart. Users can enable the `ChangeContainerStatusOnKubeletRestart` feature gate to restore the behavior where the Kubelet changes container statuses after a restart.
+
+Regarding this feature gate, the changes we will make in the kubelet codebase are as follows:
+
+  * We will retrieve the `Started` field from the container status in the Pod via the API server. After the Kubelet restarts, during the first entry into `SyncPod`, we will propagate this value to the newly generated container status.
+
```

Comment:
What about kubelet restart without API server connectivity? What's our fall back strategy for this scenario?

Reply:
without API server, there is no Pods known to kubelet. So the probes will not even start - kubelet simply doesn't know about them. Once API server is back, we will initialize with the value from API server (by that time it probably Not Ready because of lifecycle manager). And then set the value based on the actual probe result